### DISCUSSION TITLE
How should fault-handling be modeled and implemented? (+ a concern about resource space endpoints)

### DISCUSSION THREAD

**Discussion by p-linh** [sid-agrawal/OSmosis#12] [Q&A] [answered]
### The Issue
We currently pass around the fault endpoint for a PD without tracking it like our other CellulOS resources. Endpoints in general are also the only non-framework object that PDs directly handle--should we treat them as resources in the model? We need to be able to free them (through ref-counting) like the other framework resources, but it seems strange to model them as resources.

### Some points from our verbal discussion
- If we perform ref-counting on it like we do with our other resources (e.g by making it into a resource space), what does this mean in our model? 
    - A "fault endpoint component", with its own resource space,  behaves a bit differently from other CellulOS resources in that some PDs request a fault endpoint resource to listen on, while other request them to call on. Is this difference in usage sort of like permissions? 
    - Our current resource space implementation will return an endpoint to the fault endpoint, does this make sense? How will a PD who wants to listen on the fault endpoint access the actual endpoint?
    - The faulting PD will never actually directly invoke the fault endpoint, its always done by the seL4 kernel on behalf of the faulting PD - this isn't really well-represented by a hold-edge. However, the fault endpoint _does_ have to exist in the faulting PD's CSpace. 

### Should fault handlers be modelled as RDEs?
- What would the resource type on the `request` edge for this RDE? In seL4's case, each thread can only have one fault handler, which must deal with all possible faults. 
- What happens when a PD handles its own faults? There would always need to be at least one thread in a process dedicated to handling the fault, so maybe this is always represented as an RDE to another thread-PD.

**Side note about modelling of seL4 endpoints in general**
Documenting our unofficial discussion: seL4 endpoints themselves may not need to be modelled as framework resources, they seem to nicely represent the various edges of the model. We do need to track and ref-count them as if they were a resource, but this is purely an implementation detail and the fact that we need to track them seems like it's handled by the pre-existing metadata node. Does this break down anywhere?

### Related issue concerning resource space endpoints
Suppose a resource server manages multiple resource spaces, but listens on the same endpoint for all of these spaces. When one of these spaces get deleted, we can't delete the actual endpoint itself, and will need ref-counting for this endpoint as well. 

@astevins @sid-agrawal

**Comment by p-linh** [comment]
Something which I thought about as I looked back into our `request` edge definition from the paper: can our fault endpoint be synonymous with the virtual region RDE (in both implementation and model)? Then in CellulOS, the RT always handles faults for a PD unless the virtual region RDE is configured to be something else (in the case of a VMM). 

Separately, in the case of the CellulOS VMM, if the VM-PD has a virtual region RDE to the VMM, but the VMM then `request`s the RT for virtual resources to give back to the VM, does this conflict with our definition of `request` edges and RDEs? E.g. having a `request` edge to a PD that just redirects that request to another PD.

**Reply by astevins** [reply]
Are you saying it's synonymous with the VMR RDE because the faults could be page faults? Are there other types of faults for which this would not make sense?

**Reply by p-linh** [reply]
Yes, I was saying it's synonymous because I'm not sure that there are "recoverable" faults other than page faults? ~~E.g. like with capability faults, if a lookup to a cap slot fails, we don't just go and allocate a slot for it in the handler.~~ Edit: After reading some more, it seems like we can do this. Ok, so maybe faults cannot be synonymous with a regular resource-type `request` edge.

**Comment by p-linh** [comment]
Ideas from June 17th OS meeting:
- syscalls are considered request edges, so why can't faults be?
	- `request` implies there's a resource space - what is a `fault` resource space? Can't `hold` a `fault` resource either
- `request` also typically implies some operation on a resource, can faults be described in the same way?
	- E.g. in the case of `divide by 0`, maybe the argument is that this is an operation on the CPU resource
- do we need to have an edge for "providing services"? E.g. it does some computation, and what is being returned is the result of the computation?
        - E.g. ARM guest on x86 host - guest does some arch-specific syscall that has to be "translated" in the host and returned
	- fault-handler may request things from other PDs, which returns a "certificate" of having performed the necessary actions to deal with the fault
- dependency that we need to capture (faulting PD) --> (catching PD) --> (handler PD). What should be the edges between these nodes?

Conclusion for now: We could just go ahead with idea of a separate fault edge, go through some scenarios with it, and then evaluate if it can be merged with the `request` edge.
Aside: Fault radius calculations may change regardless of new edge or modelling it as a `request` edge?

**Comment by p-linh** [comment, marked as answer]
Implementation: ✅ Model: TBD
